docs/HowToFAQ/FAQLoopModifies.md

```diff
@@ -0,0 +1,90 @@
+Also, if the loop body allocates another object, that too is a change of the heap, 
+so the heap becomes an assignment target. 
+Furthermore, Dafny’s rule about a method is that a method is allowed to change the 
+state of any object in the method’s modifies clause, 
+and it is also allowed to allocate new objects and to modify their state. 
+Therefore, if your loop body calls a method, then the heap may well become an assignment target.
+
+So, then, does this mean your loop invariant needs to speak about everything in the heap whenever the heap is an assignment target of the loop?
+
+Again, no. Loops in Dafny have `modifies` clauses, just like methods do (and just like for methods, 
+newly allocated objects are implicitly included in those `modifies` clauses). 
+It would be a pain to even have to write `modifies` clauses for every loop, 
+so if a loop does not have an explicit `modifies` clause, Dafny implicitly attaches 
+the `modifies` clause of the enclosing method (or of the enclosing loop, in the case of nested loops). 
+This default behavior turns out to be right in almost all cases in practice, 
+which is why you normally don’t have to think about this issue.
+
+But when it is not, you need to be explicit about the `modifies` clause and the loop invariants. 
+In particular
+- write a `modifies` clause for the loop that includes all objects whose fields might be 
+assigned in the loop body (or listed in the modifies clauses of anything the loop body calls)
+- then write a loop invariant that includes assertions about any variable that is listed in the modifies clause
+or is an assignment target in the loop body. 
+Very typically, the invariant will give a value for a havoced variable showing its relationship to the loop index.
+
+For example, a loop that sets the elements of an array to some initial value might look like this:
+```dafny
+method init(a: array<int>) 
+  modifies a
+  ensures forall j | 0 <= j < a.Length :: a[j] == j;
+{
+  var i: int := 0;
+
+  while (i < a.Length) 
+    modifies a
+    invariant 0 <= i <= a.Length && forall j | 0 <= j < i :: a[j] == j;
+  {
+    a[i] := i;
+    i := i+1;
+  }
+}
+```
+
+Note the following points:
+- The method specifies that it might modify the elements of the array `a`.
+- The loop says that it modifies the same array. 
+In fact that modifies clause could be omitted
+because it would be inherited from the enclosing context
+- The loop also modifies `i`, but as `i` is a local variable of non-reference type, it need not be listed in the modifies clause.
+- The loop also has an invariant, which has two conjuncts
+   - One conjunct states what may be assumed about `i`. Even though `i` is not in the modifies clause 
+    it is an assignment target, so we need to say what may be assumed about it (prior to the loop test)
+   - The other conjunct says what is known about the elements of `a`, which depends on `i`, 
```

Comment:
"what may be assumed" and "what is known about" are confusing. They sound like there's no associated proof obligation. These conjuncts of the loop invariant state "what is intended to hold on loop boundaries (more precisely, what is intended to hold at the top of each iteration)". One way to modify these bullets would be to say
- One conjunct talks about the local variable `i`. ...
- The other conjunct talks about the elements of `a`. ...

Reply:
Revised